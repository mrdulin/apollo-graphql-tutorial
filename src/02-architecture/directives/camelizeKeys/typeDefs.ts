import { gql } from 'apollo-server';

const typeDefs = gql`
  directive @camelizeKeys on OBJECT
`;

export { typeDefs };
