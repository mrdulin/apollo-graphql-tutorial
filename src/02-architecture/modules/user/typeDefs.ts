import { gql } from 'apollo-server';

const typeDefs = gql`
  type User @camelizeKeys {
    userId: ID!
    userNme: String
    userEmail: String!
    userAddress: Address
  }

  extend type Query {
    userById(id: ID!): User
    users: [User]!
  }
`;

export { typeDefs };
