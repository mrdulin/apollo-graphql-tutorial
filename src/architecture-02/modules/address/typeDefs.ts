import { gql } from 'apollo-server';

const typeDefs = gql`
  type Address {
    addressId: ID!
    addressCountry: String
    addressState: String
    addressCity: String
  }

  extend type Query {
    addressById(id: ID!): Address
  }
`;

export { typeDefs };
