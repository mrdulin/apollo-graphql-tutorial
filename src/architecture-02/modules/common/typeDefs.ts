import { gql } from 'apollo-server';

const typeDefs = gql`
  type CommonResponse {
    code: Int!
    message: String!
  }
`;

export { typeDefs };
