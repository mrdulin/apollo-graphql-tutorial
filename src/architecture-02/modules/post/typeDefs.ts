import { gql } from 'apollo-server';

const typeDefs = gql`
  type Post {
    postId: ID!
    postTitle: String!
    postContent: String!
    postCreatedAt: String!
    postAuthorId: ID!
    postAuthor: User
  }

  extend type Query {
    postById(id: ID!): Post
    posts: [Post]!
  }
`;

export { typeDefs };
