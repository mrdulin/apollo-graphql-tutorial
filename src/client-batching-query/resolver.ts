import { IResolvers } from 'graphql-tools';
import faker from 'faker';

const resolvers: IResolvers = {
  Query: {
    posts: (_, __, { db }) => {
      return db.posts;
    },
    userById: (_, { id }, { db }) => {
      console.count('Query.userById');
      const user = db.users.find((u) => u.userId.toString() === id);
      return user;
    },
  },
  Post: {
    postAuthor: (parent, _, { db }) => {
      return db.users.find((user) => user.userId.toString() === parent.postAuthorId.toString());
    },
  },

  Mutation: {
    addPost: (_, { post }, { db }) => {
      return new Promise((resolve) => {
        setTimeout(() => {
          const newPost = Object.assign({}, post, { postId: faker.random.uuid() });
          db.posts.push(newPost);
          resolve({ code: 0, message: 'add post success' });
        }, 3000);
      });
    },
  },
};
export { resolvers };
