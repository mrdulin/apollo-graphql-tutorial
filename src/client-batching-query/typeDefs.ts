import { gql } from 'apollo-server';

const typeDefs = gql`
  type Post {
    postId: ID!
    postTitle: String!
    postContent: String!
    postAuthorId: ID
    postAuthor: User
  }

  input PostInput {
    postTitle: String!
    postContent: String!
    postAuthorId: ID!
  }

  type User {
    userId: ID!
    userNme: String!
    userEmail: String!
  }

  type CommonResponse {
    code: Int!
    message: String!
  }

  type Query {
    posts: [Post]!
    userById(id: ID!): User
  }

  type Mutation {
    addPost(post: PostInput): CommonResponse!
  }
`;

export { typeDefs };
