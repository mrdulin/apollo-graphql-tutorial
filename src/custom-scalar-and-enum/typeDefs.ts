import { gql } from 'apollo-server';

export const typeDefs = gql`
  enum Device {
    UNKNOWN
    DESKTOP
    HIGH_END_MOBILE
    TABLET
    CONNECTED_TV
  }

  type CampaignPerformanceReport {
    campaignNme: String!
    campaignId: ID!
    device: Device
  }

  type Query {
    campaignPerformanceReports: [CampaignPerformanceReport]!
  }
`;
