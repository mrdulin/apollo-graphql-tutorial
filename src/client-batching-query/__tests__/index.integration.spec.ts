import { createTestClient } from 'apollo-server-testing';
import { gql } from 'apollo-server';
import { ApolloServer } from 'apollo-server-express';
import { resolvers } from '../resolver';
import { typeDefs } from '../typeDefs';
import { db } from './mockedDB';

const server = new ApolloServer({ typeDefs, resolvers, context: { db } });
const testClient = createTestClient(server);

describe('client batching query', () => {
  describe('query', () => {
    it('#userById', async () => {
      const USER_BY_ID = gql`
        query userById($id: ID!) {
          userById(id: $id) {
            userNme
            userEmail
          }
        }
      `;
      const actualValue = await testClient.query({ query: USER_BY_ID, variables: { id: 1 } });
      expect(actualValue).toMatchSnapshot();
    });
  });

  describe('mutation', () => {
    it('#addPost', async () => {
      const ADD_POST = gql`
        mutation addPost($post: PostInput!) {
          addPost(post: $post) {
            code
            message
          }
        }
      `;
      const post = { postTitle: 'stackoverflow', postContent: 'stackoverflow content', postAuthorId: 1 };
      const actualValue = await testClient.mutate({ mutation: ADD_POST, variables: { post } });
      expect(actualValue).toMatchSnapshot();
    });

    it('#addPost - multiple mutations', async () => {
      const ADD_POSTS = gql`
        mutation addPost($post1: PostInput!, $post2: PostInput!) {
          addPost1: addPost(post: $post1) {
            code
            message
          }
          addPost2: addPost(post: $post2) {
            code
            message
          }
        }
      `;

      const post1 = { postTitle: 'stackoverflow', postContent: 'stackoverflow content', postAuthorId: 1 };
      const post2 = { postTitle: 'stackoverflow', postContent: 'stackoverflow content', postAuthorId: 2 };
      const actualValue = await testClient.mutate({ mutation: ADD_POSTS, variables: { post1, post2 } });
      expect(actualValue).toMatchSnapshot();
    });
  });
});
