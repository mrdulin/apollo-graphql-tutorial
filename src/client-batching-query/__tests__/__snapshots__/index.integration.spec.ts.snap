// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client batching query mutation #addPost - multiple mutations 1`] = `
Object {
  "data": Object {
    "addPost1": Object {
      "code": 0,
      "message": "add post success",
    },
    "addPost2": Object {
      "code": 0,
      "message": "add post success",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`client batching query mutation #addPost 1`] = `
Object {
  "data": Object {
    "addPost": Object {
      "code": 0,
      "message": "add post success",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`client batching query query #userById 1`] = `
Object {
  "data": Object {
    "userById": Object {
      "userEmail": "mrdulin@example.com",
      "userNme": "mrdulin",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
